[SystemZ] Global Stackprotector and associated location section

dominik-steenken opened the pull request:
This commit allows `-mstack-protector-guard=global` for `s390x`.

It also adds a new arch-specific option `-mstack-protector-guard-record`, analogous to `-mrecord-mcount`, which will cause `clang` to emit a `__stack_protector_loc` section containing all the locations in the output binary that load the stack guard address, for the purposes of later rewriting of those loads by the kernel. This new option only works together with the `global` stack protector.

In order to minimize exposure of the stack guard, both the storing of the stack guard onto the stack, and the later comparison of that value against the reference value, are handled via direct mem-to-mem instructions, those being `mvc` and `clc`.

This is achieved by introducing two new pseudo instructions, `MOVE_STACK_GUARD` and `COMPARE_STACK_GUARD`, which are inserted by the DAGCombiner after SelectionDAG construction. These pseudos stick around throughout the entire backend pipeline, and are lowered only in the AsmPrinter.

This commit also adds tests for both kinds of stack protectors (tls and global), for the proper insertion of the pseudos, the proper emission of the, `__stack_protector_loc` section, as well as the option compatibility checks for the new options.

dominik-steenken committed:
[SystemZ] Global Stackprotector and associated location section

This commit allows `-mstack-protector-guard=global` for `s390x`.

It also adds a new arch-specific option `-mstack-protector-guard-record`,
analogous to `-mrecord-mcount`, which will cause `clang` to emit a
`__stack_protector_loc` section containing all the locations in the output
binary that load the stack guard address, for the purposes of later rewriting
of those loads by the kernel. This new option only works together with the
`global` stack protector.

In order to minimize exposure of the stack guard, both the storing of the
stack guard onto the stack, and the later comparison of that value against
the reference value, are handled via direct mem-to-mem instructions, those
being `mvc` and `clc`.

This is achieved by introducing two new pseudo instructions, `MOVE_STACK_GUARD`
and `COMPARE_STACK_GUARD`, which are inserted by the DAGCombiner after
SelectionDAG construction. These pseudos stick around throughout the entire
backend pipeline, and are lowered only in the AsmPrinter.

This commit also adds tests for both kinds of stack protectors (tls and global),
for the proper insertion of the pseudos, the proper emission of the,
`__stack_protector_loc` section, as well as the option compatibility checks
for the new options.

JonPsson1 commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1830,6 +1797,28 @@ unsigned SystemZInstrInfo::getInstSizeInBytes(const MachineInstr &MI) const {
+    unsigned Size = (MI.getOpcode() == TargetOpcode::LOAD_STACK_GUARD)
+                        ? 6
+                        : 0; // lg to load value
+    if (GuardType == "global")
```
Why size 0 for the LG?

JonPsson1 commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -14,12 +14,13 @@
 #include "llvm/CodeGen/MachineInstrBuilder.h"
-#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
 #include "llvm/CodeGen/TargetLoweringObjectFileImpl.h"
```
Superfluous #include.

dominik-steenken replied:
removed

uweigand commented:
> Could we not use the same approach that the existing `LOAD_STACK_GUARD` uses in order to avoid the canary being spilled? I.e. could we not mark `LOAD_STACK_GUARD_ADDRESS` `isReMaterializable`, and do some further nudging with kill flags to ensure that the value isn't reused?

If that works reliably, I guess I'd be fine with it.  We should have test cases to demonstrate that it is isn't spilled, however.

JonPsson1 commented:
ah, I was thinking about the LOAD_STACK_GUARD_ADDRESS preg being potentially spilled.

I guess I am not quite sure about the details about when the FI is known to go out of range. Isn't that decided during PrologEpilog insertion, after all the spill slots have been created, and FrameIndices are eliminated? If it is then out of range, a vreg is allocated and used to load the address, and the vreg is handled with the RegScavenger, but that would mean that some other value is being preemptied, but the address is now in the scavenged reg. So then it wouldn't be on the stack. Or maybe this is about when there already is e.g. a huge alloca before regalloc, so the LAY is somehow produced earlier? Maybe that should be avoided somehow so it never happens early so that it always falls back on the scavenger?

I am thinking that since this is in the entry block, and in the return block(s), maybe it wouldn't be such a big loss to allocate needed regs (one for the external canary address, one for the FI address when it is out of range) as extra use operands. It would at least be simpler: one pseudo for each move/compare to be expanded late, which would also be easier to verify the location of in relation to other memory accesses. This would make this more robust against all other optimizations as well, even though those MI flags should generally be enough. But maybe the scavenger solution is better in the sense that it is only used if actually needed.

dominik-steenken committed:
Double up the StackGuard Pseudos to introduce dead defreg

The intent here is to add two pseudos with a _DAG suffix which can be slotted
into the Selection DAG in places where the instruction they are replacing does
not define a register. Then, in the custom inserter, these pseudos are replaced
with the "real" pseudos which do define an early-clobber register, which will
be assigned a physical register by regalloc, which can then be used in
ExpandPostRAPseudos to store the stack guard's address without fear that the
register might end up spilled.

dominik-steenken committed:
Implement Reviewer's Comments

This includes code formatting updates, general cleanup,
rremoving unnecessary asserts, refactoring functions for
clarity and brevity, turning `stack-protector-guard-record`
from a function attribute into a module flag, and moving
to an ISD Node approach instead of DAG combining.

For the last part, we introduce new SystemZISD nodes, CMP_STACKGUARD
and MOV_STACKGUARD, which can be inserted into the dag during isel.
These are then lowered to CMP_STACKGUARD_DAG and MOV_STACKGUARD_DAG,
then custom-inserted into their namesakes in the SystemZ namespace,
and then finally expanded post-RA to a CLC / MVC.

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1805,15 +1777,141 @@ bool SystemZInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+  // need to add 40 bytes to get the actual address of the stack guard.
+  StringRef GuardType =
+      MBB->getParent()->getFunction().getParent()->getStackProtectorGuard();
+  return (GuardType == "global") ? 0 : 40;
```
The duplicated check is a bit unfortunate.  I think it would be preferable to have a *single* `getStackGuardAddress` routine that *both* loads the address register *and* determines the appropriate offset to go with the address in that register.  [ If we end up with a single `expandStackGuardPseudo` the address load and offset determination might also simply be performed inline there.  ]

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1805,15 +1777,141 @@ bool SystemZInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+    const auto *GV = M->getNamedGlobal("__stack_chk_guard");
+    assert(GV &&
+           "could not create reference to global variable __stack_chk_guard");
+    // Ref->
```
This comment looks like a left-over?

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.td:
```diff
@@ -521,6 +521,27 @@ let SimpleBDXStore = 1, mayStore = 1 in {
+    let mayStore = 1 in {
+      // load the stack guard's address, and move the stack guard to the stack.
+      let usesCustomInserter = 1 in def MOVE_STACK_GUARD_DAG
+          : Pseudo<(outs), (ins bdaddr12only:$grdloc), []>;
```
I think we should follow the coding style everywhere else in the file and leave the "def " on the same line as the ":".

dominik-steenken replied:
done

dominik-steenken commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1056,8 +1029,7 @@ void SystemZInstrInfo::loadRegFromStackSlot(MachineBasicBlock &MBB,
   const MCInstrDesc &MCID = MI->getDesc();
-  return ((MCID.TSFlags & Flag) &&
-          isUInt<12>(MI->getOperand(2).getImm()) &&
+  return ((MCID.TSFlags & Flag) && isUInt<12>(MI->getOperand(2).getImm()) &&
```
Yes. i think `clang-format` extended to this since at some point during development i moved this function around. I'll remove the change.

dominik-steenken commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -9302,7 +9412,8 @@ SDValue SystemZTargetLowering::PerformDAGCombine(SDNode *N,
   case SystemZISD::SELECT_CCMASK: return combineSELECT_CCMASK(N, DCI);
-  case SystemZISD::GET_CCMASK:  return combineGET_CCMASK(N, DCI);
+  case SystemZISD::GET_CCMASK:
+    return combineGET_CCMASK(N, DCI);
```
will revert, as above

dominik-steenken commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1805,15 +1777,141 @@ bool SystemZInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+
+  } else if (GuardType == "global") {
+    // Obtain the global value.
+    const auto *GV = M->getNamedGlobal("__stack_chk_guard");
```
ok, via `getSDagStackGuard` i can directly get the global variable, and don't need the name at all. I'll change that.

dominik-steenken commented on clang/test/CodeGen/SystemZ/stack-guard-pseudos.c:
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -S -mllvm -stop-after=systemz-isel -stack-protector 1 -triple=s390x-ibm-linux < %s -o - | FileCheck -check-prefix=CHECK-DAGCOMBINE %s
```
OK - since we're not changing the IR that gets emitted, i will reduce this test to just the negative option check and move the rest into an LLVM check.

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.td:
```diff
@@ -521,6 +521,38 @@ let SimpleBDXStore = 1, mayStore = 1 in {
+  def LOAD_TSGA : Pseudo<(outs ADDR64:$grdaddr),
+                                            (ins), []>;
+  def LOAD_GSGA : Pseudo<(outs ADDR64:$grdaddr),
+                                               (ins), []>;
```
The short names are a bit cryptic, maybe better to have the long names.  (Could shorten a bit like `ADDR` instead of  `ADDRESS`.)

dominik-steenken replied:
I've changed the names back to long versions (`*_STACKGUARD_ADDR`).

Re: `LOAD_TGSA`: That's true i suppose. I've changed the name to `LOAD_TLS_BLOCK_ADDR`.

uweigand reviewed (commented):
I think we're getting close here.   Mostly cosmetic issues remaining ...

dominik-steenken commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -1194,6 +1194,14 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
+      CGM.getDiags().Report(diag::err_opt_not_valid_without_opt)
+          << "-mstack-protector-guard-record"
+          << "-mstack-protector-guard=global";
+    Fn->addFnAttr("mstackprotector-guard-record");
```
I have now converted this into a module flag (i.e. the function attribute `stack-protector-guard-record` was turned into a module flag of the same name).

dominik-steenken commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -3629,6 +3675,9 @@ static SDValue emitCmp(SelectionDAG &DAG, const SDLoc &DL, Comparison &C) {
                        DAG.getTargetConstant(C.ICmpType, DL, MVT::i32));
+  if (C.Opcode == SystemZISD::COMPARE_SG_DAG)
+    return DAG.getNode(SystemZISD::COMPARE_SG_DAG, DL, MVT::i32, C.Op0,
+                       DAG.getTargetConstant(C.ICmpType, DL, MVT::i32));
```
removed

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -1805,15 +1777,141 @@ bool SystemZInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+
+  } else {
+    llvm_unreachable(
+        (Twine("Unknown stack protector type \"") + GuardType + "\"")
```
But if we do want to report an error, it should probably use `report_fatal_error` - it's not a failing assertion (which would imply a programming error), but rather invalid IR input.

uweigand commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -3187,6 +3194,45 @@ static void adjustICmpTruncate(SelectionDAG &DAG, const SDLoc &DL,
+  // At this point we are sure that this is a proper compare_stack_guard
+  // case, update the opcode to reflect this.
+  C.Opcode = SystemZISD::COMPARE_STACKGUARD;
+  C.CCValid = SystemZ::CCMASK_ICMP;
```
Also, maybe set `C.Op1` to null here to indicate the COMPARE_STACKGUARD ISD only has one operand.

uweigand reviewed (commented):
I think we're getting close now.  A few remaining comments inline.

dominik-steenken commented on llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:
```diff
@@ -1026,6 +1036,70 @@ void SystemZAsmPrinter::LowerPATCHABLE_RET(const MachineInstr &MI,
+  // ear <reg>, %a1
+  EmitToStreamer(*OutStreamer,
+                 MCInstBuilder(SystemZ::EAR).addReg(Reg32).addReg(SystemZ::A1));
+  return;
```
removed

uweigand commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -3187,6 +3193,35 @@ static void adjustICmpTruncate(SelectionDAG &DAG, const SDLoc &DL,
+  // At this point we are sure that this is a proper CMP_STACKGUARD
+  // case, update the opcode to reflect this.
+  C.Opcode = SystemZISD::CMP_STACKGUARD;
+  C.CCValid = SystemZ::CCMASK_ICMP;
```
We should *not* set `CCValid` here.  The tests above ensure that it must already have been the correct value, but if it hadn't been, simply changing `CCValid` (without looking at `CCMask`) would be just wrong.

uweigand reviewed (commented):
I think this now all looks good to me; see two final remaining inline comments.

The only remaining question to decide is whether to add this with the scavenger workaround, or wait until the actual fix is committed.

sujianIBM commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.h:
Should this be removed? Thanks!

dominik-steenken replied:
Good catch, thanks.
I've removed this definition.

dominik-steenken commented:
Hello, since it seems that https://github.com/llvm/llvm-project/pull/184814 / https://github.com/llvm/llvm-project/issues/172511 might take longer to resolve than expected, can we merge this with the workaround for #172511 ?

dominik-steenken commented:
Now waiting for https://github.com/llvm/llvm-project/pull/197120.

dominik-steenken commented:
#197120 has been merged, i'll update this PR to take that into account.

dominik-steenken committed:
apply formatting rules

dominik-steenken commented:
Ok, i have rebased this onto current main, and replaced the workaround for issue #172511 with an assert on the inequality of the two registers, just in case there is a different pathway to get to that situation that wasn't considered yet.

I think this is ready for review again.

dominik-steenken commented:
@sujianIBM FYI.

uweigand reviewed (approved):
This version now LGTM - thanks for working through all the various issues!
